pkg/scheduler/plugins/nodegroup/nodegroup.go

```diff
@@ -0,0 +1,261 @@
+	"fmt"
+	"strings"
+
+	"k8s.io/klog/v2"
+
+	batch "volcano.sh/apis/pkg/apis/batch/v1alpha1"
+	sch "volcano.sh/apis/pkg/apis/scheduling/v1beta1"
+	"volcano.sh/volcano/pkg/scheduler/api"
+	"volcano.sh/volcano/pkg/scheduler/framework"
+)
+
+const (
+	// PluginName indicates name of volcano scheduler plugin.
+	PluginName       = "nodegroup"
+	NodeGroupNameKey = "volcano.sh/nodegroup-name"
+)
+
+type nodeGroupPlugin struct {
+	// Arguments given for the plugin
+	pluginArguments framework.Arguments
+}
+
+// New function returns prioritize plugin object.
+func New(arguments framework.Arguments) framework.Plugin {
+	return &nodeGroupPlugin{pluginArguments: arguments}
+}
+
+func (pp *nodeGroupPlugin) Name() string {
+	return PluginName
+}
+
+type queueGroupAffinity struct {
+	queueGroupAntiAffinityRequired  map[string][]string
+	queueGroupAntiAffinityPreferred map[string][]string
+	queueGroupAffinityRequired      map[string][]string
+	queueGroupAffinityPreferred     map[string][]string
+}
+
+func NewQueueGroupAffinity() queueGroupAffinity {
+	return queueGroupAffinity{
+		queueGroupAntiAffinityRequired:  make(map[string][]string, 0),
+		queueGroupAntiAffinityPreferred: make(map[string][]string, 0),
+		queueGroupAffinityRequired:      make(map[string][]string, 0),
+		queueGroupAffinityPreferred:     make(map[string][]string, 0),
+	}
+}
+func (q queueGroupAffinity) predicate(queue, group string) error {
+	if len(queue) == 0 {
+		return nil
+	}
+	flag := false
+	if q.queueGroupAffinityRequired != nil {
+		if groups, ok := q.queueGroupAffinityRequired[queue]; ok {
+			if contains(groups, group) {
+				flag = flag || true
+			}
+		}
+	}
+	if q.queueGroupAffinityPreferred != nil {
+		if groups, ok := q.queueGroupAffinityPreferred[queue]; ok {
+			if contains(groups, group) {
+				flag = flag || true
+			}
+		}
+	}
+	if q.queueGroupAntiAffinityPreferred != nil {
+		if groups, ok := q.queueGroupAntiAffinityPreferred[queue]; ok {
+			if contains(groups, group) {
+				flag = flag || true
+			}
+		}
+	}
+	if q.queueGroupAntiAffinityRequired != nil {
+		if groups, ok := q.queueGroupAffinityPreferred[queue]; ok {
+			if contains(groups, group) {
+				flag = flag || false
+			}
+		}
+	}
+	if !flag {
+		return errors.New("not satisfy")
+	}
+	return nil
+}
+
+func (q queueGroupAffinity) score(queue string, group string) float64 {
+	nodeScore := 0.0
+	if len(queue) == 0 {
+		return nodeScore
+	}
+	if q.queueGroupAffinityPreferred != nil {
+		if groups, ok := q.queueGroupAffinityPreferred[queue]; ok {
+			if contains(groups, group) {
+				nodeScore = 0.5
+			}
+		}
+	}
+	if q.queueGroupAffinityRequired != nil {
+		if groups, ok := q.queueGroupAffinityRequired[queue]; ok {
+			if contains(groups, group) {
+				nodeScore = 1.0
+			}
+		}
+	}
+
+	if q.queueGroupAntiAffinityPreferred != nil {
+		if groups, ok := q.queueGroupAntiAffinityPreferred[queue]; ok {
+			if contains(groups, group) {
+				nodeScore = 0.25
+			}
+		}
+	}
+
+	if q.queueGroupAntiAffinityRequired != nil {
```

Comment:
`AntiAffinityRequired` should be filtered in predict, so it should not exists in `score`